Instancer : Store prototype root attributes on instance roots

danieldresser-ie opened the pull request:
This allows for attributes to be affected by context variation.

The code is quite straightforward, though I haven't updated the tests yet.

The only tricky bit is whether this is enough of a perf cost that it should be enabled only when context variation is on.  I would much prefer this to be always this way, since that makes things simpler, and when talking with John this morning, we hypothesized that the time to generate fullAttributes shouldn't be much different, since all the attributes need to be merged anyway.

Unfortunately, in testing, I am seeing a measurable performance cost to the new code path.  A scene that generates 2 million instances with both a prototype root attribute and a point attribute ( so it requires merging ), takes about 15 seconds, and takes about 0.8 seconds longer with the new code.  This difference persists when I switch GafferSceneTest.traverseScene to call fullAttributes.

The same scene goes from taking 54 seconds to 56 seconds to do an Arnold render.  It is curious that the Scene generation time is so much higher in a render than in `gaffer stats`, and in particular that the slowdown from the new code is greater.

I guess tomorrow I should probably try to understand this performance difference ( since I'd really rather not make the code more complex by adding fast paths unless I really need to ).  I guess the most likely explanation for the slowdown may be more calls to PrototypeScope(), and more calls from there of enginePlug->getValue().  It's the same engine being retreived by every location, so it should certainly be cached, but I guess there may still be measurable cost in retrieving it in an Instancer with a huge number of lightweight instances.  I dunno, these numbers still don't quite make sense, so I'm still hoping there's some way I can justify making the code this simple.

johnhaddon reviewed (commented):
> It is curious that the Scene generation time is so much higher in a render

I don't find that all that surprising given all the work involved in converting to Arnold. 

> in particular that the slowdown from the new code is greater.

You'd have to check this, but I think prior to this we'd have been putting the prototype attribute on the capsule's procedural in Arnold once, and now we're putting it on all the instances instead. That might account for some of the extra overhead.

> I would much prefer this to be always this way, since that makes things simpler

Me too. Not just from the code maintenance perspective, but also so as not to confuse folks. User : "Where did my attribute go?". Us : "Oh, you turned on time offsets for the animation, so we moved it".

> I guess the most likely explanation for the slowdown may be more calls to PrototypeScope(), and more calls from there of enginePlug->getValue().

That seems plausible. I have noted one spot in the code where we could avoid some overhead though, so worth trying that first perhaps.

> It's the same engine being retreived by every location, so it should certainly be cached, but I guess there may still be measurable cost in retrieving it in an Instancer with a huge number of lightweight instances.

I've observed before that the cache entry for the engine can be under very heavy contention if a lot of threads are all generating instances, and there is more overhead there than we'd like. Looks like we actually get the same engine twice now, once ourselves and once implicitly in `PrototypeScope`? I wonder if it's worth having a second constructor where we can pass the engine to the scope?

I've been thinking it might be worth having a very small non-owning thread-local cache for computation results for this sort of situation. We need some sort of weak pointer support for `RefCounted` to make that possible though, as we want the global cache to be in complete control of memory usage. Weak pointers might be pretty handy in other places too...

I'll see what Murray's opinion is on the suck-up-the-overhead-or-have-two-modes question in our catch up today. But my feeling is it's reasonable to pay the small extra overhead to reduce complexity and avoid potential confusion for our users. Especially if we can claw performance back a tiny bit somewhere...

danieldresser-ie commented on src/GafferScene/Instancer.cpp:
```diff
@@ -1937,28 +1931,26 @@ void Instancer::hashBranchAttributes( const ScenePath &sourcePath, const ScenePa
 		if( e->numInstanceAttributes() )
 		{
-			return e->instanceAttributes( e->pointIndex( branchPath[2] ) );
+			CompoundObjectPtr result = prototypeAttrs->copy();
```
Fixed in place.

danieldresser-ie commented:
I've fixed the one issue you noted, and added the new signature of Protoscope, which halved the overhead in both of my tests.

I also added a 3rd commit where I try to get rid of some duplicate hashing ... this actually seems to make performance slightly worse, which doesn't make sense to me.  This 3rd commit should probably just be removed, but I thought I should upload it to see if you can spot what I'm doing wrong.

I also changed the test to match the new behaviour and updated Changes.

johnhaddon commented:
Couldn't see anything wrong with the third commit, but have pushed a rebase to remove it (and fix a couple of typos in Changes.md). Will merge once CI runs through again.
